ls1intum/Artemis
# `Quiz exercises`: Increase explanation length and provide warning if length is exceeded for quiz explanations or hints

### Checklist

### Motivation and Context


Fixes #2798 and #2799.
During development, the tooltip of the quiz question explanation was found flickering when hovering over it.




### Description

To fix #2799, the length of the fields was increased to 500. This threshold is validated by server-side integration tests.
Client-side, the invalid input flag which is already used for quizzes was extended to show a tooltip if the explanation or hint fields exceed the threshold of 500 (explanation) or 255 (hint) characters.
To fix the flickering the tooltip orientation was set to auto as it is for all other tooltips in the quiz variations.
### Steps for Testing

**Clearance for database changes by @krusche received on 09.09.2021. The changes can be deployed.**

1. Create a new Quiz and fill out all necessary parts (title) so that it can be conducted later on.
2. Add a new MC question and set a title
    2.1. Add a question hint that is longer than 255 characters
    2.2. Make sure that the invalid input flag occurs, the appropriate explanation is shown in the tooltip and the save button is 
          disabled
    2.3. Do the same with a question explanation (here up to 500 characters are possible) and make sure that another explanation 
           occurs in the tooltip
    2.4. Add some hints and explanations to at least one answer option and make sure the explanation occurs in the tooltip
    2.5 Make sure that the explanations concerning hints/explanations of answer options are only shown once, even if more than 
          one hint/explanation is invalid
3. Test the same case for question hints of DnD questions
4. In general, go wild with different combinations, also make sure that the save button is enabled again when the hints and 
explanations have an appropriate length
5. Add an explanation to the multiple-choice question if not already done and save the quiz (setting a short quiz duration is beneficial later on). Set it to visible and start it.
6. Participate in the quiz, wait for it to end and verify that the tooltip for the explanation of the quiz (not an option) doesn't flicker anymore.
7. Go back to the exercises page of the corresponding course and wait until the quiz is finished. Click on reevaluate and check that the validity checks are working there as well.
8. Do the same for a DnD quesiton

### Review Progress





- Code Review
- Manual Tests

### Test Coverage



exam-update.component.ts 100%
drag-and-drop-question-edit.component.ts 91.8%
multiple-choice-question-edit.component.ts 96.92%
quiz-exercise-detail.component.ts 98.46%

### Screenshots


This is the maximum amount of explanations that should be displayed even if more than one answer option has an invalid hint/explanation:
![image](https://user-images.githubusercontent.com/80622272/131342592-b157fcc9-df5d-4bc7-9685-8b2e90d91a65.png)

This adds an explanation to the questions:
![1ac7721aca26c98a8cb217ac848acfce](https://user-images.githubusercontent.com/64264938/131869965-4e791dda-87b0-429a-9922-d8c50b1d65b6.png)
This shows the tooltip position in the student view.
![explanation_tooltip](https://user-images.githubusercontent.com/64264938/131869968-1394d7c0-e61c-473f-9438-2792321c2e7a.jpg)
